### PR TITLE
Resolve the valgrind error shown by genericArrayInClass-otharrs-inline-iterators

### PR DESCRIPTION
**Pull request by hildeth**
This is a continuation of the effort to port the valgrind fix related to ForLoop
lowering from the string-as-rec branch to master.  The code as it stands causes an internal error to be reported when running the indicated test.  The internal error shows that the assumption that this ForLoop lowering code is never reached is not correct.  So I provided a solution that fully implements the required lowering.

DETAILS
Added the getIteratorInfoForIterator() function, mostly document the obscure way
in which we obtain the type of the underlying iterator in a ForLoop statement.

Rewrote the code that populates the map used in copying parts of a loop
construct into the various zip?() functions, hasMore(), getValue(), etc. to
clarify what it is actually doing and why.

Added code that calls down to the corresponding function in the underlying
iterator, to impelement the components of a forLoop.

Replaced "ii->foo->_this" with "my_this" where possible.

Renamed "flag" as "beforeYield", to make it more self-documenting.

Renamed getMoreBody to getValueBody inside the getValue() function, for
consistency.

Removed the genericArrayInClass-otharrs-inline-iterators test case, as it is no
longer needed.

Removed genericArrayInClass-otharrs.compopts.  It used to contain
--no-inline-iterators as a workaround for the bug that this commit fixes.  With
the fix, the workaround is no longer needed.

### CODE REVIEW

File: compiler/AST/iterator.cpp [hunk 1]
```diff
@@ -455,11 +471,24 @@ buildZip1(IteratorInfo* ii, Vec<BaseAST*>& asts, BlockStmt* singleLoop) {
...
+    VarSymbol* iterator = newTemp("forloop_iter", iteratorField->type);
+    zip1body->insertAtTail(new DefExpr(iterator));
+    CallExpr* iteratorFieldRead = new CallExpr(PRIM_GET_MEMBER_VALUE, my_this,
+                                               iteratorField);
+    zip1body->insertAtTail(new CallExpr(PRIM_MOVE, iterator, iteratorFieldRead));
+
+    // Call the contained iteratator's zip1() function. 
+    IteratorInfo* under_ii = getIteratorInfoForIterator(iterator);
+    zip1body->insertAtTail(new CallExpr(under_ii->zip1, iterator));
   } else {
```
Comment (ronawho) [inline compiler/AST/iterator.cpp]: Should remove the "ForLoops should have been replaceed in expandForLoop()." from here and buildzip3

File: compiler/AST/iterator.cpp [hunk 2]
```diff
@@ -531,15 +582,17 @@ buildZip3(IteratorInfo* ii, Vec<BaseAST*>& asts, BlockStmt* singleLoop) {
   }
 
   // Copy all non-defs in singleLoop after the yield
+  bool beforeYield = true;
```
Comment (ronawho) [inline compiler/AST/iterator.cpp]: :+1:

File: compiler/AST/iterator.cpp [hunk 3]
```diff
@@ -765,21 +937,45 @@ static void
...
-    incrBlock = cforLoop->incrBlockGet();
+      incrBlock = cforLoop->incrBlockGet();
 
-    for_alist(expr, incrBlock->body) {
-      incrBody->insertAtTail(expr->copy(&map));
+      for_alist(expr, incrBlock->body) {
+        incrBody->insertAtTail(expr->copy(&map));
+      }
     }
+    else if (singleLoop->isForLoop())
```
Comment (ronawho) [inline compiler/AST/iterator.cpp]: It's not clear to me why ForLoops need to call buildIncr.. 

The equivalent to incr should be happening for ForLoop's in one of the other buildZip* functions. Only CForLoops have an increment clause.

Comment (ronawho) [inline compiler/AST/iterator.cpp]: Just to clarify. buildIncr was explicitly added when CForLoops were added. CForLoops are basically 

`CForLoop(init, test, incr) { }`. No other loop construct has an incr clause, it all just occurs in or before the loop body and gets included in the buildZip* segments

**Comment by ronawho** [comment]
I don't really understand this patch. It seems like trying to inline a ForLoop now just results in code to call to the underlying iterator so it's not really being inlined?

I'd probably need to sit down with you to get a better idea of what's going on. OTOH, I don't know how much time I want to spend on this. It's only a for a weird case with recursive iterators that occur in a single test w/ numa.

File: compiler/AST/iterator.cpp [hunk 4]
```diff
@@ -724,36 +842,90 @@ buildHasMore(IteratorInfo* ii, BlockStmt* singleLoop) {
...
-    initBlock = cforLoop->initBlockGet();
+      initBlock = cforLoop->initBlockGet();
 
-    for_alist(expr, initBlock->body) {
-      initBody->insertAtTail(expr->copy(&map));
+      for_alist(expr, initBlock->body) {
+        initBody->insertAtTail(expr->copy(&map));
+      }
     }
+    else if (singleLoop->isForLoop())
```
Comment (ronawho) [inline compiler/AST/iterator.cpp]: Similarly to the incr clause, ForLoops shouldn't have a init clause..

File: compiler/AST/iterator.cpp (hunk 3 above)
Comment (hildeth) [inline compiler/AST/iterator.cpp]: The ForLoop being replaced would not give rise to an incr clause, but the underlying iterator might.  I didn't want to distinguish how the underlying iterator is rendered, so I just avoided the question by fully implementing the interface.  If someone can prove that the init and incr wrappers can be empty, then the corresponding if (singleLoop->isForLoop()) clauses can be removed from the corresponding buildInit() and buildIncr() routines.
I think I have a counterexample, though I also am reaching the point where much additional effort on this single failing test case seems unwarranted.

**Comment by hildeth** [comment]
Along those lines -- It is interesting that you pointed out my solution doesn't really inline the implementation of the ForLoop.  That is true.  I was entirely focused on obtaining the correct behavior, so I overlooked the original intention of inlining the loop implementation as well.
I think this could be done easily, by taking the contents of the underlying iterator's implementation function and inlining it at the point of the call.  Nothing special needs to be done here, so we can probably just set the FLAG_INLINE flag on those iterator implementation functions (zip?,  init, incr, hasMore and hasValue) and let the inlining pass handle thos downstream.
In terms of performance, I can probably argue that even that makes no difference.  The C compiler will see a number of functions that look very simple, such as::

   init1(IC1* ic1_) {
     IC2* ic2_ = ic1_->F1__iterator;
     init2(ic2_);
   }

which it is likely to inline, even without an invitation.  If that is true, then inlining them in the code produced by Chapel would be wasted effort.

**Comment by hildeth** [comment]
@ronawho : I'll check with you tomorrow, to see how you want to proceed on this.

**Comment by ronawho** [comment]
I agree that we don't want to spend much more time on this.

Let's chat about this for a couple minute tomorrow. I think this is really a recursive iterator lowering bug. I know we do __not__ have codegen for the ForLoop ast elements (codegen has an INT_FATAL.) I'm guessing this ForLoop must then be getting lowered into a CForLoop, but I don't really know what's going on and I really, really do not want to have to wrap my head around recursive iterator lowering. 

I don't think the test passing is a good indication of a "correct" implementation though. With the big caveat that I'm not sure how recursive iterator lowering works, check this out: https://github.com/chapel-lang/chapel/compare/master...ronawho:recursive-iter-lowering-fix

**Pull request merged**